File: 0401_0500/0482/0482.py
with open("input.txt", "r") as f, open("output.txt", "w") as q:
    n = int(f.readline())
    i, u, s = 1, 1, 0
    while s + u < n:
        s += u
        i += 1
        u += 1 if i < 10 else 2 if i < 100 else 3
    n -= s
    if n >= 190:
        n -= 190
        n = 1 if n % 3 == 0 else n // 30 if n % 3 == 1 else n % 30 // 3
    elif n >= 10:
        n -= 10
        n = n % 20 // 2 if n % 2 else n // 20 + 1
    q.write(str(n))
        

File: 0401_0500/0482/0482.c
#include <stdio.h>

int main()
{
	FILE *f = fopen("input.txt", "r"), *q = fopen("output.txt", "w");
	int n, s = 0;
	char c;

	fscanf(f, "%d", &n);
	for (int i = 1, u = 1; s + u < n; ++i)
		s += u, u += i + 1< 10 ? 1 : i + 1 < 100 ? 2 : 3;
	n -= s;
	if (n < 10)
		c = n + '0';
	else if (n < 190)
	{
		n -= 10;
		c = n % 2 ? n % 20 / 2 + '0' : n / 20 + '1';
	}
	else
	{
		n -= 190;
		c = n % 3 == 0 ? '1' : n % 3 == 1 ? n / 30 + '0' : n % 30 / 3 + '0';
	}
	putc(c, q);
	return 0;
}
